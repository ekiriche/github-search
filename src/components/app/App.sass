.repository-list
  max-width: 700px
  margin-left: auto
  margin-right: auto

.pagination-wrapper
  display: flex
  justify-content: center
  padding: 12px 0