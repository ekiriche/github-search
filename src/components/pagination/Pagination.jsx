import React from "react";
import {Pagination as MaterialPagination} from '@material-ui/lab';

function Pagination(props) {
    return (
        <MaterialPagination
            page={props.activePage}
            onChange={props.onChange}
            count={Math.ceil(props.totalItems / props.itemsPerPage)}
            shape="rounded"
            color="primary"
        />
    );
}

Pagination.defaultProps = {
    itemsPerPage: 30
};

export default Pagination;