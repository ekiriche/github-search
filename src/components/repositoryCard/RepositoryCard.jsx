import React from 'react';
import StarIcon from '@material-ui/icons/Star';
import * as githubColors from '../utils/githubLanguageColors';
import './RepositoryCard.sass';

const RepositoryCard = (props) => {
    return (
        <div className="card-container">
            <a className="card-name" href={props.nameLink}>{props.name}</a>
            {props.description &&
                <p className="card-description">
                    {props.description.length > 117 ? `${props.description.slice(0, 118)}...` : props.description}
                </p>
            }
            <div className="bottom-wrapper">
                <a className="stars-container" href={`${props.nameLink}/stargazers`}>
                    <StarIcon/>
                    {props.stars}
                </a>
                {props.language &&
                    <div className="language-container">
                        <span
                            className="language-color"
                            style={{
                                backgroundColor: githubColors.default[props.language] ?
                                    githubColors.default[props.language].color : null
                            }}
                        />
                        {props.language}
                    </div>
                }
                {props.license && props.license !== "Other" &&
                    <div className="license">
                        {props.license}
                    </div>
                }
                <div className="updatedAt">
                    {`Updated ${props.updatedAt}`}
                </div>
            </div>
        </div>
    );
};

export default RepositoryCard;