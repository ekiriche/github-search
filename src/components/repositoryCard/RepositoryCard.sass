.card-container
  padding: 12px
  max-width: 700px

.card-name
  color: #0366d6
  text-decoration: none
  &:hover
    text-decoration: underline

.bottom-wrapper
  display: flex
  align-items: center
  font-size: 12px

.card-description
  margin: 4px 0
  font-size: 14px

.stars-container
  color: #586069
  cursor: pointer
  text-decoration: none
  display: flex
  align-items: center
  &:hover
    color: #0366d6

.star-icon
  font-size: 16px

.language-container, .license, .stars-container
  margin-right: 14px

.language-color
  position: relative
  top: 1px
  display: inline-block
  width: 12px
  height: 12px
  border-radius: 50%
  margin-right: 4px