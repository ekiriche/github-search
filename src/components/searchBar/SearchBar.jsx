import React from "react";
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import InputBase from '@material-ui/core/InputBase';
import SearchIcon from '@material-ui/icons/Search';
import GitHubIcon from '@material-ui/icons/GitHub';
import { createStyles, makeStyles, fade } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) =>
    createStyles({
        appRoot: {
            backgroundColor: "#24292e"
        },
        searchContainer: {
            position: 'relative',
            borderRadius: theme.shape.borderRadius,
            backgroundColor: fade(theme.palette.common.white, 0.15),
            marginRight: theme.spacing(2),
            marginLeft: 0,
            width: '100%'
        },
        searchIcon: {
            padding: theme.spacing(0, 2),
            height: '100%',
            position: 'absolute',
            pointerEvents: 'none',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
        },
        inputRoot: {
            color: 'inherit',
        },
        inputInput: {
            fontSize: theme.typography.fontSize,
            padding: theme.spacing(1, 1, 1, 0),
            paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
            transition: theme.transitions.create('width'),
            width: '100%',
            [theme.breakpoints.up('sm')]: {
                width: '20ch',
                '&:focus': {
                    width: '26ch',
                },
            },
        },
        container: {
            display: "flex",
            alignItems: "center"
        },
        githubIcon: {
            cursor: "pointer",
            fontSize: '32px',
            marginRight: theme.spacing(2),
            '&:hover': {
                color: "hsla(0, 0%, 100%, .7)",
                textDecoration: "none"
            },
        }
    })
);

const SearchBar = (props) => {
    const classes = useStyles();

    const onSubmit = (e) => {
        e.preventDefault();
        props.onSubmit();
    };

    return (
        <AppBar
            position="static"
            classes={{
                root: classes.appRoot
            }}
        >
            <Toolbar>
                <form onSubmit={(e) => onSubmit(e)}>
                    <div className={classes.container}>
                        <GitHubIcon className={classes.githubIcon} onClick={props.onIconClick}/>
                        <div className={classes.searchContainer}>
                            <div className={classes.searchIcon}>
                                <SearchIcon />
                            </div>
                            <InputBase
                                placeholder={props.placeholder}
                                value={props.value}
                                onChange={props.onChange}
                                classes={{
                                    root: classes.inputRoot,
                                    input: classes.inputInput,
                                }}
                            />
                        </div>
                    </div>
                </form>
            </Toolbar>
        </AppBar>
    );
};

export default SearchBar;