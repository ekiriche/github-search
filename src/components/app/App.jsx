import React, {useState} from 'react';
import './App.sass';
import SearchBar from "../searchBar/SearchBar";
import {AxiosRequests} from "../utils";
import RepositoryCard from "../repositoryCard/RepositoryCard";
import Pagination from "../pagination/Pagination";
import moment from "moment";
import "./App.sass";

const axiosRequests = new AxiosRequests();

const App = () => {
    const [searchValue, setSearchValue] = useState("");
    const [searchResult, setSearchResult] = useState(null);
    const [currentPage, setCurrentPage] = useState(1);

     const searchRepos = (page) => {
         setSearchResult(null);
         if (window.sessionStorage.getItem(`${searchValue}-${page}`)) {
             setSearchResult(JSON.parse(window.sessionStorage.getItem(`${searchValue}-${page}`)));
         } else {
             axiosRequests.getReposRequest(searchValue, page, 30).then((response) => {
                 setSearchResult(response.data);
                 window.sessionStorage.setItem(`${searchValue}-${page}`, JSON.stringify(response.data));
             }).catch((error) => {
                 console.log(error);
             });
         }
     };

    const onPageChange = (page) => {
        searchRepos(page);
        setCurrentPage(page);
    };

    return (
      <div>
          <SearchBar
              value={searchValue}
              placeholder="Search..."
              onChange={(e) => setSearchValue(e.target.value)}
              onSubmit={() => onPageChange(1)}
              onIconClick={() => axiosRequests.getReposRequestAbort()}
          />
          <div className="repository-list">
              {searchResult && <h3>{`${Math.min(searchResult.total_count, 1000)} repository results`}</h3>}
              {searchResult && searchResult.items.map((item, index) => (
                  <RepositoryCard
                      key={index}
                      name={item.full_name}
                      nameLink={item.html_url}
                      description={item.description}
                      stars={item.stargazers_count}
                      language={item.language}
                      license={item.license && item.license.name}
                      updatedAt={moment(item.updated_at).fromNow()}
                  />
                  ))}
          </div>
          <div className="pagination-wrapper">
              {searchResult && searchResult.total_count ?
                  <Pagination
                      activePage={currentPage}
                      onChange={onPageChange}
                      totalItems={Math.min(searchResult.total_count, 1000)}
                  /> : null
              }
          </div>
      </div>
    );
};

export default App;
