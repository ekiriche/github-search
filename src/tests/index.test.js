import React from 'react';
import renderer from 'react-test-renderer';
import App from '../components/app/App';
import Pagination from "../components/pagination/Pagination";
import RepositoryCard from "../components/repositoryCard/RepositoryCard";
import SearchBar from "../components/searchBar/SearchBar";

test('expect <App /> to match snapshot', () => {
    const component = renderer.create(
        <App />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('expect <Pagination /> to match snapshot', () => {
    const component = renderer.create(
        <Pagination activePage={1} totalItems={450} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('expect <RepositoryCard /> to match snapshot', () => {
    const component = renderer.create(
        <RepositoryCard />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('expect <SearchBar /> to match snapshot', () => {
    const component = renderer.create(
        <SearchBar />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});