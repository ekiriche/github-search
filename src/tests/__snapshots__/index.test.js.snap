// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect <App /> to match snapshot 1`] = `
<div>
  <header
    className="MuiPaper-root MuiAppBar-root makeStyles-appRoot-1 MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="makeStyles-container-6"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root makeStyles-githubIcon-7"
            focusable="false"
            onClick={[Function]}
            viewBox="0 0 24 24"
          >
            <path
              d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
            />
          </svg>
          <div
            className="makeStyles-searchContainer-2"
          >
            <div
              className="makeStyles-searchIcon-3"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <div
              className="MuiInputBase-root makeStyles-inputRoot-4"
              onClick={[Function]}
            >
              <input
                className="MuiInputBase-input makeStyles-inputInput-5"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Search..."
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  </header>
  <div
    className="repository-list"
  />
  <div
    className="pagination-wrapper"
  />
</div>
`;

exports[`expect <Pagination /> to match snapshot 1`] = `
<nav
  aria-label="pagination navigation"
  className="MuiPagination-root"
>
  <ul
    className="MuiPagination-ul"
  >
    <li>
      <button
        aria-label="Go to previous page"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiPaginationItem-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </button>
    </li>
    <li>
      <button
        aria-current="true"
        aria-label="page 1"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-selected"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        1
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 2"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        2
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 3"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        3
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 4"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        4
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 5"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        5
      </button>
    </li>
    <li>
      <div
        className="MuiPaginationItem-root MuiPaginationItem-ellipsis"
      >
        â€¦
      </div>
    </li>
    <li>
      <button
        aria-label="Go to page 15"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        15
      </button>
    </li>
    <li>
      <button
        aria-label="Go to next page"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiPaginationItem-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`expect <RepositoryCard /> to match snapshot 1`] = `
<div
  className="card-container"
>
  <a
    className="card-name"
  />
  <div
    className="bottom-wrapper"
  >
    <a
      className="stars-container"
      href="undefined/stargazers"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </a>
    <div
      className="updatedAt"
    >
      Updated undefined
    </div>
  </div>
</div>
`;

exports[`expect <SearchBar /> to match snapshot 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root makeStyles-appRoot-1 MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="makeStyles-container-6"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root makeStyles-githubIcon-7"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
          />
        </svg>
        <div
          className="makeStyles-searchContainer-2"
        >
          <div
            className="makeStyles-searchIcon-3"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            className="MuiInputBase-root makeStyles-inputRoot-4"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input makeStyles-inputInput-5"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</header>
`;
